### PR TITLE
Add Jest 29 support, remove Jest 28 support, and remove Node.js 12 support

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Ref: https://github.com/stylelint/stylelint/pull/6308

> Is there anything in the PR that needs further explanation?

https://jestjs.io/blog/2022/08/25/jest-29

### CODE REVIEW

**Comment by JounQin** [comment]
If we want a consistent format, we need to set `snapshotFormat` specifically.

**Comment by ybiquitous** [comment]
I don't think we need to keep the snapshot format.

**Comment by JounQin** [comment]
> I don't think we need to keep the snapshot format.

`jest: "^25.3.0 || ^26.0.1 || ^27.0.1 || ^28.0.2 || ^29.0.1"` means different `snapshotFormat` by default.

Maybe we can just use

```diff
+ snapshotFormat: {
+   escapeString: false,
+   printBasicPrototype: false
+ }
```

**Comment by ybiquitous** [comment]
The `jest-preset-stylelint` package simply provides a test helper not related to snapshots. So, I don't think we need to add extra settings; it's better to follow the default settings of each Jest version.

Users of `jest-preset-stylelint` can configure `snapshotFormat` by themselves or update their snapshots to fit Jest 29.

**Comment by JounQin** [comment]
Interesting, I would imagine `jest-preset-stylelint` results consistent behavior/snapshots for all `jest`(peer) versions.

File: package.json [hunk 1]
```diff
@@ -69,9 +69,9 @@
     "typescript": "^4.7.3"
   },
   "peerDependencies": {
-    "jest": "^25.3.0 || ^26.0.1 || ^27.0.1 || ^28.0.2"
+    "jest": "^25.3.0 || ^26.0.1 || ^27.0.1 || ^28.0.2 || ^29.0.1"
   },
   "engines": {
-    "node": "^12.20.0 || ^14.13.1 || >=16.0.0"
+    "node": "^14.15.0 || >=16.0.0"
```
Comment (ybiquitous) [inline package.json:75]:
```console
$ npm v jest@29.0.1 engines
{ node: '^14.15.0 || ^16.10.0 || >=18.0.0' }
```

**Comment by ybiquitous** [comment]
Jest 29 removed the support for Node.js 12. So, we also need to remove it.

The next version of `jest-preset-stylelint` should be `6.0.0` (breaking change).

**Review by JounQin — APPROVED** [review]

**Review by jeddy3 — APPROVED** [review]
Can this be released independently of a major release of Stylelint or do they go hand-in-hand?

**Comment by JounQin** [comment]
> Can this be released independently of a major release of Stylelint or do they go hand-in-hand?

Personally, I think one major version for all changes is fine.

**Comment by ybiquitous** [comment]
> Can this be released independently of a major release of Stylelint or do they go hand-in-hand?

Yes. But, we cannot update `jest` and `jest-preset-stylelint` in the stylelint/stylelint repo until dropping Node.js 12 support (see <https://github.com/stylelint/stylelint/issues/6311>).

For now, only authors of Stylelint plugins that have already dropped Node.js 12 support will benefit from this update.

> Personally, I think one major version for all changes is fine.

Of course, that would make things simple, but not required.

**Comment by jeddy3** [comment]
> Yes. But, we cannot update jest and jest-preset-stylelint in the stylelint/stylelint repo until dropping Node.js 12 support (see https://github.com/stylelint/stylelint/issues/6311).

Good stuff. Let's merge and release a new major version of this package. We can update the Jest dependencies in `stylelint` when we release our next major version of Stylelint.

**Comment by ybiquitous** [comment]
[6.0.0](https://github.com/stylelint/jest-preset-stylelint/releases/tag/6.0.0) released! 🎉

File: package.json (hunk 1 above)
Comment (Mouvedia) [reply package.json]: Wait so only LTS versions?
i.e. 15.x.x is not supported?

**Comment by Mouvedia** [comment]
You forgot to update the tsconfig.json.
i.e. `"ES2020"`
see https://github.com/stylelint/stylelint/issues/6311#issuecomment-1234851501

File: package.json (hunk 1 above)
Comment (ybiquitous) [reply package.json]: We have run linting only on Node.js LTS. For example, see the [CI job log](https://github.com/stylelint/stylelint/runs/8188711914?check_suite_focus=true).

Node.js v15 [reached End-of-Life](https://github.com/nodejs/Release/#end-of-life-releases).

**Comment by ybiquitous** [comment]
@Mouvedia Thanks for reminding me. I've created PR #53.

**Pull request merged**